Autoloads with UIDs

KnifeXRage opened the pull request:
Closes: [#111720](https://github.com/godotengine/godot/issues/111720)

Autoloads can now use UIDs for `Scripts` and `Scenes`.

<img width="401" height="130" alt="Screenshot 2025-10-30 154258" src="https://github.com/user-attachments/assets/c3fa911c-a3af-4b2e-9a5d-c198acb94256" />

Autoload Using UIDs:

[Autoload UIDs](https://github.com/user-attachments/assets/76d9bad5-80f4-438d-9d1a-177f9c75c8ff)

When Autoloads files moved Externally:

[Moving Autoload Files Externally](https://github.com/user-attachments/assets/161c6f31-31e9-456b-8515-a08832f90b08)

KnifeXRage committed:
Autoloads with UIDs

KoBeWi commented on core/io/resource_uid.cpp:
```diff
@@ -37,6 +37,10 @@
 
+#ifdef TOOLS_ENABLED
+#include "editor/settings/editor_autoload_settings.h"
+#endif
```
Don't include editor code in core 💀
Why do you even need this?

KnifeXRage replied:
I added this to clear `filesystem_cache`, if a unrecognised UID found in project. Because if you move autoload file externally and not it's UID file then it causes too much unrecognised UID errors and even if you move that uid to it's actual path still error will not solve until unless you manually delete the `filesystem_cache` in ".godot/editor/". This will automatically delete if it Recognises Unrecognised UID and fix that 😅.

KoBeWi reviewed (commented):
While the idea is fine, there are some unnecessary changes here. What is the file cache thing for? If you want to cleanup outdated autoloads or something then it should be a separate PR.

KnifeXRage commented:
As you said i have done all requested changes. but for `packed_scene.cpp` it's a required to convert UIDs to real path if using scene Autoloads. so i updated it to `path = ResourceUID::ensure_path(path);`. 

And i also removed auto cache clear feature. now if _cache_ cause problems to load Autoloads, now we need to **manually** clear it.

KoBeWi commented on editor/settings/editor_autoload_settings.cpp:
```diff
@@ -427,6 +427,7 @@ Node *EditorAutoloadSettings::_create_autoload(const String &p_path) {
 
 void EditorAutoloadSettings::init_autoloads() {
 	for (AutoloadInfo &info : autoload_cache) {
+		info.path = ResourceUID::ensure_path(info.path); // Convert UID to actual Path (If Possible)
```
What exactly needs a path to load properly? The only problematic part is `reload_from_file()`, which is still handled separately.

KnifeXRage replied:
But the path we need to provide in `create_autoload` function's argument is file path not UID

KoBeWi commented on scene/resources/packed_scene.cpp:
```diff
@@ -2451,6 +2451,8 @@ void PackedScene::clear() {
 
 void PackedScene::reload_from_file() {
 	String path = get_path();
+	path = ResourceUID::ensure_path(path); //If it's a resource uid, then convert to real path.
```
What `get_path()`?
There are no errors or anything.

KoBeWi replied:
Yes.

KnifeXRage replied:
Done :)

Repiteo commented:
Thanks! Congratulations on your first merged contribution! 🎉
